[add]コメント編集＆削除＆いいね　コメント通知　

ryuzo111 opened the pull request:
【説明文】
コメント編集機能と削除機能といいね機能の実装
コメント時の投稿ユーザーにメール機能を通知
(細かな修正　:　ヘッダーのアプリ名をクリックすると、投稿一覧ページに遷移、goodsテーブルのカラム名がpost_idからcomment_idに修正)

【コマンド】
(新たにgoodsテーブルのfactoryを作成したので以下コマンドを行うことにより、goodsのテストデータを格納することができます)
php artisan migrate:refresh --seed
mailableクラスを作成したので、.envでmailドライバの設定をお願いいたします。

【期待される結果】
・自分のコメントには編集＆削除できるようになる。
・他のユーザーのコメントにはいいねをすることができる。
・いいね済みのコメントのいいねを取りけしすることができる。
・コメントをした場合は、その投稿をしたユーザーにメール通知が行く
・ヘッダーの「laravel」をおすと投稿一覧に遷移する


お忙しいところ恐縮ですが、ご確認いただけますと幸いです。
また、改善点がございましたら、ご教示お願いいたします。

ryuzo111 committed:
[add]悩み投稿の詳細ページ&コメント機能

2つの機能を追加しました。

ryuzo111 committed:
[update]detailのviewを修正

ログインしていないとコメントできないという文言を追加

ryuzo111 committed:
[update]detailのviewを変更

コメント一覧をログイン有無を問わずに閲覧可能に変更

ryuzo111 committed:
[update]viewのインデントを修正,timezoneをAsiaに変更,Modelのreturnをtrueに変更

レビューいただいたコメントをもとに上記のような内容に修正

ryuzo111 committed:
Merge branch 'master' into feature-comment-delete-edit-good

ryuzo111 committed:
[add]コメント編集機能とコメント削除機能といいね機能、コメント時の通知機能

コメント周りを完成しました。

komo15 commented:
・deleteすると、「page you are looking for could not be found」になってしまいます。
（削除はできてます）

・仕様の確認ですが、コメント時のメール文「コメントの詳細はログインしてご確認ください」とありますが、
実際はメール上でコメント全文が見れる仕様かと思います。
こちらは、別途編集予定でしょうか？

ryuzo111 committed:
[fix]コメントを削除したときの処理とメールのメッセージを変更

ryuzo111 committed:
[fix]コメント時のメール文を編集

rhapsody0919 commented:
>mailableクラスを作成したので、.envでmailドライバの設定をお願いいたします。
これは、どう設定したら良いか書いていただきたいです。

ryuzo111 commented:
.envファイルの以下の内容を追加していただけたらと思います。

MAIL_DRIVER=smtp
MAIL_HOST=smtp.mail.yahoo.co.jp
MAIL_PORT=465
MAIL_USERNAME=
MAIL_PASSWORD=
MAIL_ENCRYPTION=ssl
MAIL_FROM_ADDRESS=
MAIL_FROM_NAME=

例)gmailだと以下のようになります。
MAIL_DRIVER=smtp
MAIL_HOST=smtp.gmail.com
MAIL_PORT=587
MAIL_USERNAME=<送信元gmailのアドレス>   #gmailの場合、USERNAMEはFROM_ADDRESSと同じ
MAIL_PASSWORD=<先程のアプリパスワード>
MAIL_ENCRYPTION=tls
MAIL_FROM_ADDRESS=<送信元gmailのアドレス>
MAIL_FROM_NAME=ECSiteSample   #メールの件名なのでご自由に

https://qiita.com/koru1893/items/e30d19ac97eac59b1e19

不足点がございましたら、ご教示お願いいたします。

ryuzo111 commented:
申し訳ございません。
開発ルールを破っている記述があります。
修正いたします。

ryuzo111 committed:
[fix]開発ルールに沿ったコードに変更

Requestと関数名の名前を変更

ryuzo111 commented:
開発ルールに沿ったコードに変更いたしました。
何度も申し訳ございません。
確認お願い致します。

rhapsody0919 commented on app/Http/Controllers/CommentController.php:
```diff
@@ -4,19 +4,65 @@
     }
-    
-    public function comment(CommentRequest $request)
+
```
投稿や編集など、成功or失敗でそれぞれフラッシュメッセージの作成をしたいですね。

rhapsody0919 commented on app/Http/Controllers/CommentController.php:
```diff
@@ -4,19 +4,65 @@
+        $comment_id = $this->comment->saveComment($request);
+        $comment_data = $this->comment->getCommentById($comment_id);
+        Mail::send(new CommentNotification($comment_data));
+        return redirect()->route('post.detail', ['post_id' => $request->input('post_id')]);
```
メールが送られてきていないのは僕だけですかね？(汗)

rhapsody0919 replied:
すいません!
完全に勘違いしておりました!

rhapsody0919 commented on routes/web.php:
```diff
@@ -36,7 +36,13 @@
 
     Route::post('/delete/{id}', 'PostController@delete')->name('post.delete');
+
     Route::post('/comment', 'CommentController@comment')->name('post.comment');
```
prefixがcommentでグループ化した方が良いですね!
Route::group(['prefix' => 'post', ...
のなかで、さらに、グループ化ってできるんでしたっけ？

ryuzo111 replied:
グループ化の中にグループ化できました！

ryuzo111 committed:
[update]routeの変更とflash_messageの追加

レビューいただいた内容をもとに機能の追加

ryuzo111 commented:
レビューをもとに再度pushしました！
【変更点】
・コメントを追加、編集、削除、いいね、いいね取り消し時にflash_messageを表示
・web.phpのrouteを修正(commentの部分をRouteにまとめる)

ご確認お願いいたします。

komo15 commented:
良いと思います！グループの中にグループ作れるんですね。勉強になります。

rhapsody0919 commented:
マイグレーションファイルを編集したら、
コマンドの記入お願いします。。。笑

rhapsody0919 commented on app/Http/Controllers/CommentController.php:
```diff
@@ -4,19 +4,70 @@
+        $comment_id = $this->comment->saveComment($request);
+        $comment_data = $this->comment->getCommentById($comment_id);
+        Mail::send(new CommentNotification($comment_data));
+        session()->flash('flash_message', 'コメントを投稿しました');
```
セッションメッセージが表示されないですね。。。

ryuzo111 replied:
私のPCだと正しく表示されます、、

ryuzo111 replied:
調べたら、以下コマンドを実行すると表示されるそうでした。
コマンド漏れ失礼しました。

 php artisan view:clear

rhapsody0919 replied:
まだ、表示されないですね、、、

rhapsody0919 replied:
失礼しました。PRを更新したところ表示されました。

rhapsody0919 commented on app/Http/Controllers/CommentController.php:
```diff
@@ -4,19 +4,70 @@
+
+    public function delete(Request $request)
+    {
+        $comment_data = $this->comment->getCommentById($request->input('comment_id'));
```
$comment_dataを見た時は、正直commentの中身が入っているかと思いました。$commentでいいと思います。冗長です。

rhapsody0919 commented on app/Http/Requests/SaveCommentRule.php:
```diff
@@ -28,7 +28,8 @@ public function rules()
+    public function messages()
+    {
         return [
             'post_id.required' => 'エラーです',
```
>エラーです。
だけでは、少し心許ないかと思います。
後で文言の訂正は行う必要がありますよね？
お任せします。

rhapsody0919 commented on app/Models/Comment.php:
```diff
@@ -17,12 +17,36 @@ public function post()
+    }
+    public function editComment($comment_data)
+    {
+        $comment = $this->getCommentById($comment_data->comment_id);
```
ここで$commentを使いたいがために、$comment_dataを用いているのですね。
$target_commentとかの方が一般的な気がします!

rhapsody0919 commented on resources/views/post/detail.blade.php:
```diff
@@ -2,6 +2,10 @@
 @section('content')
 
+@if (session('flash_message'))
+    <div class="flash_message" style="color:red">{{ session('flash_message') }}</div>
```
これは、app.blade.phpに書いてあげたほうが共通化できますね。

あと、セッションメッセージが表示されないです、、、

ryuzo111 committed:
[fix]変数名の変更とバリデーションエラーのメッセージを変更

controllerとmodelの$comment_dataの変数名を$commentと$target_commentに変更、バリデーションエラーのメッセージを変更

ryuzo111 commented:
【変更点】
・modelとControllerの変数名を編集
$comment_data → $comment
(コメント編集に関わる処理の場合はviewにわたす変数を$target_commentに変更)
・RequestのEditCommentRuleとSaveCommentRuleのバリデーションメッサージを変更
・view側のsession表示をdetail.blade.phpからapp.blade.phpに転記して共通化しました。
(session表示は次のコマンドで表示されるか試していただけないでしょうか？「php artisan view:clear」)

ご確認お願いいたします。

rhapsody0919 commented on app/Http/Controllers/CommentController.php:
```diff
@@ -4,19 +4,70 @@
+    public function deleteGood(Request $request)
+    {
+        $this->good->deleteGoodByCommentId($request->input('comment_id'));
+        session()->flash('flash_message', 'コメントのいいねを取り消しました');
```
できたら、flash_messageではなく、"success" or "error"で使い分けましょうか!

rhapsody0919 commented on app/Models/Comment.php:
```diff
@@ -17,12 +17,36 @@ public function post()
+        return $this->id;
+    }
+
+    public function deleteCommentById($comment_id)
```
直す必要はないですが、
$comment_idは、冗長ですね。
開発ルールで、
「主キーで更新するなら主キーのみ」とあることから、
$idと書くだけで分かります。

deleteCommentByPostId($post_id)というように外部キー$post_idを使用するなら明記する必要があります!

ryuzo111 replied:
そうなのですね！

そのあたりの考え方が自分にはないので、大変勉強になります。

rhapsody0919 commented on app/Models/Good.php:
```diff
@@ -0,0 +1,32 @@
+            $this->where('comment_id', $comment_id)->where('user_id', Auth::id())->delete();
+            return true;
+        } else {
+            abort(403, '権限がありません');
```
基本的には、「コントローラがviewに表示するものを受け渡す」というのがMVCモデルの考え方なので、modelに'権限がありません'と書くのはあまりよろしくないです。
viewに受け渡すものはコントローラ内で処理しましょう。

rhapsody0919 commented:
すいません、前回の指摘漏れもありました。
できたら、
・flash_messageではなく、"success" or "error"で使い分けましょうか!
・コントローラからviewへの受け渡し
を終わったら自分でマージまでしてもらって大丈夫です!

ryuzo111 committed:
コンフリクト解消

ryuzo111 committed:
[fix]sessionメッセージを表示するclassの変更&いいね取り消し時のviewの受け渡しをcontrollerに変更

sessionメッセージを表示するクラス名をsuccessとerrorに変更、モデルにてviewで表示する方法からconrollerからviewに受け渡す方法に変更
